README.md
# recordsapi

Simple data-parsing CLI/API


## Usage

### As CLI

**Show help screen:**

```bash
$ lein run -- --help
```


**Import/view data:**

```bash
$ lein run -- --file $INPUT_PATH --view 1
```

Where `view` can be one of:

- **1**: sort data by email address descending, then last name ascending
- **2**: sort data by birthdate ascending
- **3**: sort data by last name descending


Data to import must be text files with the following format:

| Last name | First name | Email address | Favorite color | Birthdate  |
| ---       | ---        | ---           | ---            | ---        |
| Smith     | Sarah      | s@hotmail.com | mauve          | 1980-08-22 |

Record separators can be either "`, `" (comma+space), "` | `" (space+pipe+space), or "` `" (space). The separator must be consistent throughout a given file.

Sample data in all supported formats is available in `./resources`:

| Filename | Format          |
| ---      | ---             |
| data.csv | Comma-separated |
| data.psv | Pipe-separated  |
| data.ssv | Space-separated |


Dates must be ISO-8601 format (`yyyy-MM-dd`).


### As web server

**Start the server:**

```bash
$ lein run -- --server
```

As with the CLI, you may optionally load a file here via `--file`.


**Consume the API:**

| Action | Shell command |
| ---    | ---      |
| Get records sorted by email | `curl http://localhost:3030/records/email` |
| Get records sorted by birthdate | `curl http://localhost:3030/records/birthdate` |
| Get records sorted by last name | `curl http://localhost:3030/records/name` |
| Add a record (comma-separated values) | `curl -H 'Content-Type: application/json' -d '{"line":"foo, bar, foobar@example.com, red, 1973-03-22"}' http://localhost:3030/records` |
| Add a record (pipe-separated values) | `curl -H 'Content-Type: application/json' -d '{"line":"baz \| quux \| qb@example.com \| jade \| 1986-10-31"}' http://localhost:3030/records` |
| Add a record (space-separated values) | `curl -H 'Content-Type: application/json' -d '{"line":"schmoe joe joe@me.com violet 1953-12-13"}' http://localhost:3030/records` |

*Note*: [HTTPie](https://httpie.io/) is a more user-friendly CLI web client than cURL, and will display JSON/response codes in a more pleasing and informative way.


### Run tests

```bash
$ lein test
```


## License

Proprietary

Copyright © 2021 Jeff Kayser


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2021-02-15
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2021-02-15
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/recordsapi/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/recordsapi/compare/0.1.0...0.1.1


test/recordsapi/cli_test.clj
(ns recordsapi.cli-test
  (:require [clojure.test :refer :all]
            [clojure.string :as s]
            [clojure.tools.cli :refer [parse-opts]]
            [recordsapi.cli :as cli]))

(deftest cli-options
  (testing "CLI options spec is a vec of vecs"
    ; Just a very basic structure check -- tools.cli already validates strictly via spec
    (is (vector? cli/cli-options))
    (is (every? true? (map vector? cli/cli-options)))))

(deftest usage
  (testing "usage screen is a string"
    (is (string? (cli/usage [])))))

(deftest error-msg
  (testing "error message screen includes errors"
    (let [errors ["FOO!" "and BAR!"]]
      (is (map #(s/includes? (cli/error-msg errors) %) errors)))))

(defn fails-validation? [args]
  (let [bad (cli/validate-args args)]
    (is (contains? bad :exit-message))
    (is (not= true (:ok? bad)))))

(deftest validate-args
  (testing "args are validated as expected"
    (let [help (cli/validate-args ["--help"])]
      (is (contains? help :exit-message))
      (is (= true (:ok? help))))

    (fails-validation? ["--asdf"])
    (fails-validation? ["--view 0"])
    (fails-validation? ["--view 4"])
    (fails-validation? ["--port 0"])
    (fails-validation? ["--port 65536"])
    (fails-validation? ["--port foo"])
    (fails-validation? ["--file /some/place/that/probably/does/not/exist.txt"])

    (let [good (:options (cli/validate-args ["-s" "-v2" "--file" "/dev/null" "--port" "65535"]))]
      (is (= true (:server good)))
      (is (= 2 (:view good)))
      (is (= (clojure.java.io/file "/dev/null") (:file good)))
      (is (= 65535 (:port good))))))


test/recordsapi/web_test.clj
(ns recordsapi.web-test
  (:require [clojure.test :refer :all]
            [ring.util.response :refer [response content-type]]
            [ring.mock.request :refer [request]]
            [recordsapi.data :refer [db]]
            [recordsapi.web :as web])
  (:import [java.time LocalDate]))

(def foo {:last "bar" :first "foo" :color "green" :email "me@hrm.com" :birthdate (LocalDate/of 1970 1 1)})
(def bar {:last "asdf" :first "qwerty" :color "red" :email "asdf@hrm.com" :birthdate (LocalDate/of 1966 12 31)})

(defn init-test-data [test-fn]
  (reset! db [foo bar])
  (test-fn))

(use-fixtures :once init-test-data)

(deftest wrap-content-type
  (testing "sets Content-Type header to JSON"
    (let [handler (-> (constantly (response "foo"))
                      web/wrap-content-type-json)
          resp (handler (request :get "/"))]
      (is (= resp {:status  200
                   :headers {"Content-Type" "application/json"}
                   :body    "foo"})))))

(deftest wrap-body-json
  (testing "returns body as JSON"
    (let [handler (-> (constantly (response {:foo "bar"
                                             :b 123}))
                      web/wrap-body-json)
          resp (handler (request :get "/"))]
      (is (= resp {:status  200
                   :headers {}
                   :body    "{\"foo\":\"bar\",\"b\":123}"})))))

(deftest add-record
  (testing "adds record"
    (is (= {:status 201
            :body   {:message "Record successfully added"
                     :success true}})
        (web/add-record "{\"line\": \"a,b,c,d,1000-01-01}"))

    (is (= {:status 400
            :body   {:message "Failed to process specified record"
                     :success false}})
        (web/add-record "{\"foo\": \"another bar?"))))

(deftest get-records-by
  (testing "gets properly sorted records"
    (is (= [bar foo] (get-in (web/get-records-by :email) [:body :records])))
    (is (= [bar foo] (get-in (web/get-records-by :birthdate) [:body :records])))
    (is (= [bar foo] (get-in (web/get-records-by :last) [:body :records])))))

(deftest routes
  (testing "routing and response are correct"
    (let [resp {:status 200 :headers {} :body {:message "<placeholder>"
                                               :success true
                                               :records (list bar foo)}}]
      ; get by email
      (is
        (= (assoc-in resp [:body :message] "Got records sorted by email")
           (web/routes {:uri "/records/email" :request-method :get})))

      ; get by birthdate
      (is
        (= (assoc-in resp [:body :message] "Got records sorted by birthdate")
           (web/routes {:uri "/records/birthdate" :request-method :get})))

      ; get by last name
      (is
        (= (assoc-in resp [:body :message] "Got records sorted by last")
           (web/routes {:uri "/records/name" :request-method :get})))

      ; bad URL
      (is
        (= {:status 404 :headers {} :body {:message "Not found"
                                           :success false}}
           (web/routes {:uri "/foo" :request-method :get})))
        )))


test/recordsapi/core_test.clj
(ns recordsapi.core-test
  (:require [clojure.test :refer :all]
            [clojure.string :as s]
            [recordsapi.core :as core])
  (:import [java.time LocalDate]))

(deftest format-record
  (testing "formats birthdate properly, changes nothing else"
    (is (= {:f         "foo"
            :birthdate "1/2/2000"}
           (core/format-record {:f         "foo"
                                :birthdate (LocalDate/of 2000 1 2)})))
    (is (= {:b         "bar"
            :birthdate "1/22/2000"}
           (core/format-record {:b         "bar"
                                :birthdate (LocalDate/of 2000 1 22)})))))

(deftest format-output
  (testing "tables are formatted properly"
    (is (= "| a   | asdf  | baz |"
           (core/format-output
             "\n  |   a |  asdf | baz |   \t \n\r"))
        )))

(deftest print-records
  (testing "records are output per spec"
    (is (=
          (with-out-str
            (core/print-records
              [{:a "A" :b "B!" :birthdate (LocalDate/of 1986 5 13)}]))
          (s/join "\n" ["| :a | :b | :birthdate |"
                        "|----+----+------------|"
                        "| A  | B! | 5/13/1986  |\n"])))))


src/recordsapi/cli.clj
(ns recordsapi.cli
  (:require [clojure.java.io :as io]
            [clojure.string :as s]
            [clojure.tools.cli :refer [parse-opts]]))

(def cli-options
  [["-f" "--file INPUT" "Input file path"
    :parse-fn io/file
    :validate [#(.exists %) "File not found"]]
   ["-v" "--view TYPE" "Output view (ignored in \"web server\" mode). Sorts by: 1=email descending (default), last name; 2=birthdate; 3=last name descending"
    :parse-fn #(Integer/parseInt %)
    :validate [#(<= 1 % 3)]
    :default 1]
   ["-s" "--server" "Run API web server"
    :default false]
   ["-p" "--port PORT" "Web server port"
    :default 3030
    :parse-fn #(Integer/parseInt %)
    :validate [#(<= 1 % 65535) "Must be between 1 and 65535, inclusive"]]
   ["-h" "--help" "Show this help screen"]])

(defn usage [options-summary]
  (->> ["Records API"
        ""
        "Usage: lein run -- [options]"
        ""
        "Options:"
        options-summary
        ""]
       (s/join "\n")))

(defn error-msg [errors]
  (str "Errors occurred while parsing your command:\n\n"
       (s/join "\n" errors)))

(defn validate-args [args]
  (let [{:keys [options arguments errors summary]}
        (parse-opts args cli-options)]
    (cond
      ; Help requested; show usage
      (:help options) {:exit-message (usage summary) :ok? true}

      ; Problem parsing; show errors
      errors {:exit-message (error-msg errors)}

      ; Args parsed correctly
      :else {:options options})))

(defn exit [status msg]
  (println msg)
  (System/exit status))


(comment
  (parse-opts ["--file" "/tmp/foo.txt" "--view" "2"] cli-options)

  (validate-args ["--file /tmp/foo.txt" "--server" "-p1234"])
  )


src/recordsapi/core.clj
(ns recordsapi.core
  (:require [clojure.pprint :as pp]
            [clojure.string :as s]
            [recordsapi.cli :as cli]
            [recordsapi.data :as data :refer [db]]
            [recordsapi.web :as web])
  (:import [java.time.format DateTimeFormatter])
  (:gen-class))

; Output date format
(def date-format (DateTimeFormatter/ofPattern "M/d/yyyy"))

(defn format-record
  "Format record per specification"
  [record]
  (into record {:birthdate (.format (:birthdate record) date-format)}))

(defn format-output
  "Trim and left-justifies ASCII table"
  [str]
  (-> str
      s/trim
      (s/replace #"\| (\s*)([^\s]+)" "| $2$1")))

(defn print-records
  "Output records in final, user-friendly format"
  [records]
  (->> records
       (map format-record)
       pp/print-table
       with-out-str
       format-output
       println))

(defn -main [& args]
  (let [{:keys [options exit-message ok?]} (cli/validate-args args)]
    (when exit-message
      (cli/exit (if ok? 0 1) exit-message))
    (when (:file options)
      (data/load-data-from-file! (:file options)))
    (if (:server options)
      (web/start-server (:port options))
      (print-records (data/get-records (:view options))))))


(comment

  (load-data! (clojure.java.io/resource "data.csv"))
  (count @db)
  @db
  (reset! db [])

  (defn output-line [line]
    (flatten [(butlast (vals line)) (.format (last (vals line)) date-format)])
    )

  (map (comp load-data! clojure.java.io/resource) ["data.csv" "data.psv" "data.ssv"])


  )


src/recordsapi/data.clj
(ns recordsapi.data
  (:require [clojure.java.io :as io]
            [clojure.string :as s]
            [clojure.data.csv :as csv])
  (:import [java.time LocalDate]))

; Input file field order
(def fields [:last :first :email :color :birthdate])

; Our record "database"; a vector of maps
(defonce db (atom []))

(defn read-first-line
  "Read the first line of `path`"
  [path]
  (with-open [reader (io/reader path)]
    (first (line-seq reader))))

(defn detect-sep
  "Attempt to detect the record separator character used in the specified `line`"
  [line]
  (condp #(s/includes? %2 %1) line
    " | " \|
    ", " \,
    " " \space
    (throw (Exception. "Unknown record separator"))))

(defn parse-cols
  "Parse string cols into desired storage data type"
  [cols]
  (flatten [(map s/trim (butlast cols))
            (-> cols last s/trim LocalDate/parse)]))

(defn parse-string
  "Parse input CSV from `str`, autodetecting the record separator"
  [str]
  (let [first-line (s/split str #"\r?\n" 2)
        sep (-> first-line detect-sep)]
    (map parse-cols (csv/read-csv str :separator sep))))

(defn mapify
  "Convert records into a map matching the expected input spec"
  [cols]
  (zipmap fields cols))

(defn read-data
  "Get fully parsed record maps from `path`"
  [path]
  (map mapify (parse-string (slurp path))))

(defn add-record!
  "Add record to 'database' atom"
  [rec]
  (swap! db conj rec))

(defn load-data-from-string!
  [str]
  (doall
    (map add-record!
         (map mapify (parse-string str)))))

(defn load-data-from-file!
  "Load data from path into 'database'"
  [path]
  (load-data-from-string! (slurp path)))

(defn get-records
  "Get records sorted by specified mode"
  [view-mode]
  (condp = view-mode
    1 (sort-by :email #(compare %2 %1) (sort-by :last @db)) ; Email descending, then last name
    2 (sort-by :birthdate @db)                              ; Birthdate
    3 (sort-by :last #(compare %2 %1) @db)                  ; Last name descending
    (throw (Exception. (str "Invalid view option: '" view-mode "'")))))


(comment

  (map mapify (parse-string (slurp (io/resource "data.csv"))))
  (parse-string (slurp (io/resource "data.psv")))
  (parse-string (slurp (io/resource "data.ssv")))

  (def space "B A ab@foo.net red 2001-02-03\nY X x.y@bar.org blue 2010-01-31\n")
  (load-data-from-string! space)
  (load-data-from-file! (io/resource "data.csv"))


  )


src/recordsapi/web.clj
(ns recordsapi.web
  (:require [compojure.core :refer [defroutes context GET POST]]
            [compojure.route :as route]
            [jsonista.core :as json]
            [ring.adapter.jetty :as jet]
            [recordsapi.data :as data :refer [db]]))

(defn wrap-content-type-json [handler]
  (fn [request]
    (let [response (handler request)]
      (assoc-in response [:headers "Content-Type"] "application/json"))))

(defn wrap-body-json [handler]
  (fn [request]
    (let [response (handler request)]
      (assoc response :body (json/write-value-as-string (:body response))))))

(defn add-record
  [raw]
  (try
    (let [line (-> raw json/read-value (get "line"))]
      (data/load-data-from-string! line))
    {:status 201
     :body   {:message "Record successfully added"
              :success true}}
    (catch Exception e
      {:status 400
       :body   {:message "Failed to process specified record"
                :success false}})))

(defn get-records-by [sort-field]
  {:status 200
   :body   {:message (str "Got records sorted by " (name sort-field))
            :success true
            :records (sort-by sort-field @db)}})

(defroutes routes
  (context "/records" []
    (POST "/" {body :body} (add-record (slurp body)))
    (GET "/email" [] (fn [_] (get-records-by :email)))
    (GET "/birthdate" [] (fn [_] (get-records-by :birthdate)))
    (GET "/name" [] (fn [_] (get-records-by :last))))
    (route/not-found {:status 404
                      :body   {:message "Not found"
                               :success false}}))

(def app
  (-> routes
      wrap-body-json
      wrap-content-type-json))

(defn start-server [port]
  (jet/run-jetty app {:port  port
                      :join? false}))



(comment

  (wrap-content-type-json {:uri "/"} {:body ""})

  (json/write-value-as-string {:a "foo"})
  (json/read-value "{\"a\": \"foo\"}")

  )
